Add minAssuranceForResolution to register.

mandyvenables opened the pull request:
This PR adds `minAssuranceForResolution` to the `register` endpoint, and stores `minAssuranceForResolution` in the `registration` and `tokenBatch` tables. It also adds the  `levelOfAssurance ` property to the `resolve` endpoint, and checks that against `minAssuranceForResolution` and throws an error if `'Minimum Level of Assurance not met.'`.

dmitrizagidulin commented on lib/tokens.js:
```diff
@@ -342,6 +346,15 @@ export async function resolve({requester, token}) {
     const {tokenBatch} = await _getBatch({id: batchId});
     const {internalId} = tokenBatch;
 
+    if(levelOfAssurance <= tokenBatch.minAssuranceForResolution) {
```
This should be `<` instead of `<=`, no?
(Since an initial registration on the Onboarding site will have `minAssuranceForResolution: 2`. And when it's verified by the clerk, the token resolution request will have: `levelOfAssurance: 2`, and it should pass.)

mandyvenables replied:
Correct, and updated.

mattcollier commented on lib/tokens.js:
```diff
@@ -249,11 +249,14 @@ export async function registerDocumentAndCreate(
  */
 export async function create(
-  {internalId, attributes = new Uint8Array(), tokenCount} = {}) {
+  {internalId, attributes = new Uint8Array(), tokenCount,
```
please use the same indent pattern as the `register` API on L110.

mattcollier commented on lib/tokens.js:
```diff
@@ -288,7 +291,8 @@ export async function create(
     // 2. Get an open batch for creating tokens.
     const {tokenBatch, startIndex, claimedTokenCount} = await _getOpenBatch(
-      {internalId, batchVersion, tokenCount: target});
+      {internalId, batchVersion, tokenCount: target,
+        minAssuranceForResolution});
```
Opening brace moves up to L293 and closing brace moves to new line.

mattcollier commented on lib/tokens.js:
```diff
@@ -342,6 +346,15 @@ export async function resolve({requester, token}) {
 
+    if(levelOfAssurance < tokenBatch.minAssuranceForResolution) {
+      throw new BedrockError(
+        'Minimum Level of Assurance not met.',
```
```suggestion
        'Minimum level of assurance not met.',
```

mattcollier commented on lib/tokens.js:
```diff
@@ -342,6 +346,15 @@ export async function resolve({requester, token}) {
+      throw new BedrockError(
+        'Minimum Level of Assurance not met.',
+        'NotAllowedError', {
+          public: true,
```
i think it's helpful to include the details for the expected value and the actual value in cases like this.  So maybe in addition to the existing details:
```
{
  levelOfAssurance,
  minAssuranceForResolution: tokenBatch.minAssuranceForResolution
}
```

mattcollier commented on lib/tokens.js:
```diff
@@ -665,7 +678,8 @@ async function _getKek({hmac, version, salt}) {
 }
 
-async function _createBatch({internalId, batchVersion, tokenCount = 0}) {
+async function _createBatch({internalId, batchVersion, tokenCount = 0,
```
please use the same indent pattern as the register API on L110.

mattcollier commented on lib/tokens.js:
```diff
@@ -679,7 +693,8 @@ async function _createBatch({internalId, batchVersion, tokenCount = 0}) {
   const remainingTokenCount = Math.max(0, batchTokenCount - tokenCount);
   if(remainingTokenCount === 0) {
-    return _insertBatch({id, internalId, batchVersion, tokenCount});
+    return _insertBatch({id, internalId, batchVersion, tokenCount,
```
leave opening curly brace and move all params to next line.

mattcollier commented on lib/tokens.js:
```diff
@@ -708,18 +723,21 @@ async function _createBatch({internalId, batchVersion, tokenCount = 0}) {
   circumstances, the non-ideal cases may still not occur. */
   const [record] = await Promise.all([
-    _insertBatch({id, internalId, batchVersion, tokenCount}),
+    _insertBatch({id, internalId, batchVersion, tokenCount,
```
leave opening curly brace and move all params to next line.

mattcollier commented on lib/tokens.js:
```diff
@@ -784,7 +802,8 @@ async function _getUnfilledBatch({internalId, batchVersion}) {
 }
 
-async function _insertBatch({id, internalId, batchVersion, tokenCount}) {
+async function _insertBatch({id, internalId, batchVersion, tokenCount,
+  minAssuranceForResolution}) {
```
please use the same indent pattern as the register API on L110.

mattcollier reviewed (approved):
TY.  Style nits.  Approving.

dlongley commented on lib/documents.js:
```diff
@@ -106,8 +108,9 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
-  internalId, externalId, document, recipients, recipientChain, ttl, creator,
-  tokenizer, externalIdHash, documentHash, creatorHash, newRegistration
+  internalId, externalId, document, recipients, recipientChain,
+  minAssuranceForResolution, ttl, creator, tokenizer, externalIdHash,
```
Is the intent for this to be a breaking change? If not, we need a default. Reading back over the design doc indicates we should pick `2` as the default. We'll want to be careful how we set this default, however, only doing so when the registration is new and not a refresh of an existing one. In other words, the presence of `minAssuranceForResolution` indicates that we want to set its value to what was passed vs. taking one of two paths:

1. When registering a new one, use the value passed OR the default if one wasn't passed.
2. When refreshing an existing registration, use the value passed OR do NOT update `minAssuranceForResolution`.

This brings me to the fact that the `_refresh` path also needs consideration here to allow us to update `minAssuranceForResolution` on an already registered document.

That refresh is currently quite fast -- which we don't want to change. However, since we are storing `minAssuranceForResolution` in two places now -- both in token batches and in registered documents, we need to figure out how to keep them in sync or decide that we don't want to/need to.

I'm going to take this off to a design doc to discuss what we want to do.

dlongley commented on lib/documents.js:
```diff
@@ -140,10 +144,18 @@ export async function register({
+        externalIdHash, documentHash, ttl, creatorHash,
+        minAssuranceForResolution
+      })) {
+      const {registration} =
+        await _getRegistrationRecord({externalIdHash, documentHash});
+      await updateBatch({
```
So this is the hottest code path. We need the flag here we discussed that determines whether or signals whether or not batches should be updated so that we can only very rarely execute these two calls. Also, it's potentially more than one batch that gets updated, I believe.

dlongley commented on lib/documents.js:
```diff
@@ -186,6 +199,44 @@ export async function register({
     }
   }
 }
+/**
```
```suggestion

/**
```

dlongley commented on lib/documents.js:
```diff
@@ -186,6 +199,44 @@ export async function register({
+} = {}) {
+  const tokenizer = await tokenizers.getCurrent();
+
+  const {externalIdHash, documentHash, creatorHash} =
+      await _tokenizeRegistration({tokenizer, externalId, document, creator});
+
+  const {registration} =
+    await _getRegistrationRecord({externalIdHash, documentHash});
+
+  await _refresh({
+    externalIdHash, documentHash, ttl, creatorHash, minAssuranceForResolution,
+  });
+
+  await updateBatch({
+    internalId: registration.internalId, minAssuranceForResolution
+  });
```
We need a large comment block above all of these steps that details how it won't create atomicity issues -- and where the potential pitfalls are, if any.

dlongley replied:
I'm also concerned about what happens when this function is called and the registration has expired. The comments should explain how that isn't an issue with details for why.

mandyvenables replied:
This has been refactored, and a comment has been added to indicate if it gets to that point it cannot be expired since it will have just been refreshed in the `if`.

dlongley commented on lib/documents.js:
```diff
@@ -373,7 +428,8 @@ async function _refresh({externalIdHash, documentHash, ttl, creatorHash}) {
       'meta.updated': now,
-      'registration.expires': new Date(now + ttl)
+      'registration.expires': new Date(now + ttl),
+      'registration.minAssuranceForResolution': minAssuranceForResolution
```
Did we decide that we always want to set this? Or do we only want to set it if it is passed in?

mandyvenables replied:
Updated to check if it's passed in.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +465,32 @@ export async function resolveToInternalId({token}) {
+ *
+ * @returns {boolean} Returns true if update occurred.
+ */
+export async function updateBatch({internalId, minAssuranceForResolution}) {
```
```suggestion
export async function updateBatch({internalId, minAssuranceForResolution} = {}) {
```

Also, can't there be multiple batches open for `internalId`?

dlongley replied:
Is it ok for closed batches to be updated?

mandyvenables replied:
Updated to update multiple batches.

What exactly do you mean by "closed batches"? I have added a check for `remainingTokenCount: {$gt:0}` and the batches will automatically be deleted when they expire, is there another definition for `closed`?

mattcollier commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+## 5.0.0 - 2021-03-xx
+
+### Added
+- **BREAKING**: Implement `levelOfAssurance` in resolve endpoint.
+- **BREAKING**: Implement `minAssuranceForResolution` in create endpoint.
```
I assume this is breaking because these are now `required` fields.  If that's the case, please update the verbiage to something like:  `Implement new required field `levelOfAssurance` in resolve endpoint.`

mattcollier commented on lib/documents.js:
```diff
@@ -140,10 +146,20 @@ export async function register({
+        externalIdHash, documentHash, ttl, creatorHash,
+        minAssuranceForResolution
+      })) {
+      const {registration} =
+        await _getRegistrationRecord({externalIdHash, documentHash});
```
```suggestion
      const {registration} = await _getRegistrationRecord(
        {externalIdHash, documentHash});
```

mattcollier commented on lib/tokens.js:
```diff
@@ -449,6 +465,32 @@ export async function resolveToInternalId({token}) {
+ * @param {number} options.minAssuranceForResolution - Minimum level of
+ *   assurance required for token resolution.
+ *
+ * @returns {boolean} Returns true if update occurred.
```
```suggestion
 * @returns {Promise<boolean>} Returns true if update occurred.
```

mattcollier reviewed (approved):
Minor nits for my part, approving.

dlongley commented on lib/documents.js:
```diff
@@ -140,10 +146,20 @@ export async function register({
+      })) {
+      const {registration} =
+        await _getRegistrationRecord({externalIdHash, documentHash});
+      if(updateTokenBatch) {
+        await updateBatch({
```
There might be more than one batch, right? Shouldn't this be pluralized and be changed to handle updating all open batches?

mandyvenables replied:
Updated.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +465,41 @@ export async function resolveToInternalId({token}) {
 }
 
+/**
+ * Updates a token batch based off of `internalId`.
```
```suggestion
 * Sets the `minAssuranceForResolution` on the token batches associated `internalId`.
```

dlongley commented on lib/documents.js:
```diff
@@ -11,6 +11,7 @@ import canonicalize from 'canonicalize';
 import {Cipher} from 'minimal-cipher';
 import {promisify} from 'util';
 import {createKeyResolver} from './keyResolver.js';
+import {updateBatches} from './tokens.js';
```
This creates a circular dependency. We need to fix it. I think the way to fix it -- and to make the error I mention down below make a bit more sense is to change the `updateBatches` flag to a flag that indicates that we require that the registration NOT be new. We can use the existing `newRegistration` flag to handle it having three possible values: `true`, `false`, or `undefined`. When we want to update an existing registration, we set it to `false` and throw if there is no existing registration. Then, if an existing registration is successfully updated, we can call `updateBatches` (which we may also want to rename because it is specifically changing the `minAssuranceForResolution`).

```suggestion
```

dlongley commented on lib/documents.js:
```diff
@@ -102,12 +105,15 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
  *   (tokenized) creator.
  * @param {boolean} [options.newRegistration] - Optionally specify whether or
  *   not a registration is expected to be new.
+ * @param {boolean} [options.updateTokenBatches] - Specify whether or
+ *   not to update token batches.
```
```suggestion
```

dlongley commented on lib/documents.js:
```diff
@@ -102,12 +105,15 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
  *   (tokenized) creator.
  * @param {boolean} [options.newRegistration] - Optionally specify whether or
  *   not a registration is expected to be new.
```
```suggestion
 *   not a registration is expected to be new; if this is set to `false`,
 *   and the registration does not exist, then an error will be thrown.
```

dlongley commented on lib/documents.js:
```diff
@@ -102,12 +105,15 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
-  tokenizer, externalIdHash, documentHash, creatorHash, newRegistration
+  internalId, externalId, document, recipients, recipientChain,
+  minAssuranceForResolution = 1, ttl, creator, tokenizer, externalIdHash,
+  documentHash, creatorHash, newRegistration, updateTokenBatches
```
```suggestion
  documentHash, creatorHash, newRegistration
```

dlongley commented on lib/documents.js:
```diff
@@ -140,10 +146,35 @@ export async function register({
     // 3. Attempt to refresh the expiration for the document if it is not
     //    known to be a new registration.
     if(!newRegistration &&
```
```suggestion
    if(newRegistration !== true &&
```

dlongley commented on lib/documents.js:
```diff
@@ -140,10 +146,35 @@ export async function register({
+        externalIdHash, documentHash, ttl, creatorHash,
+        minAssuranceForResolution
+      })) {
+      const {registration} = await _getRegistrationRecord({
+        externalIdHash, documentHash
+      });
+      // this will only update the tokenBatches if there is a valid
+      // registration. It will not be expired since it was just refreshed.
+      if(updateTokenBatches) {
+        await updateBatches({
+          internalId: registration.internalId, minAssuranceForResolution
+        });
+      }
```
```suggestion
```

dlongley commented on lib/documents.js:
```diff
@@ -140,10 +146,35 @@ export async function register({
       // doc already registered and now refreshed, no need to encrypt; return
       // existing record
-      return _getRegistrationRecord({externalIdHash, documentHash});
+      return {registration};
```
```suggestion
      return _getRegistrationRecord({externalIdHash, documentHash});
```

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +465,41 @@ export async function resolveToInternalId({token}) {
+ *
+ * @returns {Promise<boolean>} Returns true if update occurred.
+ */
+export async function updateBatches({internalId, minAssuranceForResolution}) {
```
I recommend we rename this to `setMinAssuranceForResolution` to keep it specific to its job. Depending on our security analysis elsewhere, we may find we need `raiseMinAssuranceForResolution` and `lowerMinAssuranceForResolution` (or similar) functions.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +465,41 @@ export async function resolveToInternalId({token}) {
+    'tokenBatch.internalId': internalId,
+    'tokenBatch.remainingTokenCount': {$gt: 0}
+  };
+  const records = await collection.find(query).toArray();
+  let result;
+  for(const record of records) {
+    const updateQuery = {
+      'tokenBatch.id': record.tokenBatch.id
+    };
+    const $set = {
+      'meta.updated': Date.now(),
+      'tokenBatch.minAssuranceForResolution': minAssuranceForResolution
+    };
+
+    result = await collection.updateOne(
+      updateQuery, {$set}, database.writeOptions);
+  }
```
This isn't the right way to update multiple records. Instead of finding and retrieving them -- use `updateMany` and do a single call.

mandyvenables commented:
@dlongley I updated to use `newRegistration` flag instead of previous `updateTokenBatches`, changed the function name to `setMinAssuranceForResolution` added a function to access a registration from the `externalId`.

dlongley commented on lib/documents.js:
```diff
@@ -252,6 +282,30 @@ export async function getRegistration({internalId} = {}) {
+ *
+ * @returns {Promise<object>} The registration record.
+ */
+export async function getRegistrationbyExternalId({externalId}) {
```
This method is problematic. There can be multiple registration records for a given `externalId`, they are only unique for a combination of `externalIdHash` and `documentHash`. If we *have* to expose a call to get a registration, it will need both the external ID and the document. But I suspect we don't even want to expose such a thing -- because it likely has race conditions with its usage.

mandyvenables replied:
Removed.

mandyvenables commented on lib/documents.js:
```diff
@@ -139,11 +144,35 @@ export async function register({
+      // registration. It will not be expired since it was just refreshed.
+      await setMinAssuranceForResolution({
+        internalId: registration.internalId, minAssuranceForResolution
+      });
```
Updated.

mandyvenables commented:
@dlongley I have removed `minAssuranceForResolution` from the `_refresh` call and into separate functions that independently adjust the `minAssuranceForResolution` for the registration and token batches. I think this is better because it gives us finer control of `minAssuranceForResolution` instead of resetting it each time, and the routes that call it to update registration and token batches will occur less frequently than other calls and should not negatively impact performance.

dlongley commented:
@mandyvenables, Ok, if it's true that adjusting `minAssuranceForResolution` can be moved entirely out of hot paths such that we have more control over it (and there are no related atomicity issues by doing so), that does sound like the way to go. We'll need to analyze the new approach by stepping through the flows (and document this) and checking it against the code (and then, eventually experimentally confirm it's not a hot path, but that's later). Thanks!

mandyvenables commented on lib/tokens.js:
```diff
@@ -449,6 +467,62 @@ export async function resolveToInternalId({token}) {
+  const $set = {
+    'meta.updated': Date.now(),
+    'tokenBatch.remainingTokenCount': 0
+  };
```
Thanks for pointing this out. I had tested this on another repo and was able to get the desired behavior, however when tested directly it did not provide the expected behavior.

I have updated the `invalidateTokenBatch` to `expireTokenBatch`, and when it is called it set the expiration date of the token batch to the current Date. I have also added the parameter that a tokenBatch must not be expired to the `_getBatch` function, which will return a `NotFoundError`. It is possible to add an `allowExpired` flag, however the comment block prior to the `resolve` function notes a token will not resolve if it is expired, so I do not believe that is necessary.

dlongley replied:
You'll want to check to see what happens if the expiration is set to the current date. Does it trigger mongodb's background worker to remove it from the database? If so -- is that ok for auditing purposes or do the token batches need to stick around? I suspect they need to stick around. It may be that we actually need another flag to mark a batch as invalid/unresolvable vs. expiring it early.

mandyvenables replied:
I have set a new flag `invalid: true` on tokenBatches that have been invalidated. I also set the `remainingTokenCount` to 0 so no new tokens are created from that batch.

dlongley replied:
Ok, thanks, that sounds good, I'll review again soon as I can.

mandyvenables commented on lib/documents.js:
```diff
@@ -249,6 +266,37 @@ export async function getRegistration({internalId} = {}) {
+  internalId, minAssuranceForResolution
+}) {
+  const query = {
+    'registration.internalId': internalId
```
Updated.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +469,64 @@ export async function resolveToInternalId({token}) {
 
+/**
+ * Sets the `minAssuranceForResolution` on the token batches associated
+ *   `internalId`.
```
We don't indent function descriptions.

```suggestion
 * `internalId`.
```

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +469,64 @@ export async function resolveToInternalId({token}) {
+}
+
+/**
+ * Invalidates the token batch by marking it as invalid, but leaves
+ *   the record intact so previously resolved tokens are trackable. The
+ *   remaining token count is set to 0 so no new tokens can be created in the
+ *   batch.
```
```suggestion
 * the record intact so previously resolved tokens are trackable. The
 * remaining token count is set to 0 so no new tokens can be issued from the
 * batch.
```

dlongley replied:
```suggestion
 * Invalidates all token batches associated with the given `internalId`
 * by marking them as invalid, preventing future resolution of any issued
 * tokens. The token batch records remain intact so previously resolved
 * tokens are linkable, but the remaining token count is set to 0 so no new
 * tokens can be issued from the batch.
```

dlongley reviewed (approved):
This is looking good! I left some comments with documentation changes that we need, but I'm approving. I think we should wait to merge until we've confirmed that we're able to handle our external use cases properly (we need to make sure that the atomicity -- or lack thereof -- of the exposed APIs ensures those cases are covered). But if so, we're good here once the doc changes are in.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +469,64 @@ export async function resolveToInternalId({token}) {
+ */
+export async function invalidateTokenBatch({internalId}) {
+  const query = {
+    'tokenBatch.internalId': internalId
```
Hmm, I think I spoke too soon. We don't have an index on this and we can't add one. We're going to have to figure out what to do here.

dlongley commented on lib/tokens.js:
```diff
@@ -449,6 +469,64 @@ export async function resolveToInternalId({token}) {
+  internalId, minAssuranceForResolution
+}) {
+  const query = {
+    'tokenBatch.internalId': internalId,
```
Similar problem here -- we can't do a query on `tokenBatch.internalId` -- we have no such index. We should have a meeting that includes @mattcollier to discuss options here.

dlongley replied:
@mandyvenables, so the decision from the meeting was to add a context for `tokenBatch.internalId` to `tokenization-tokenBatch` -- even though it will not include the shard key. Since we don't expect this code to execute in the hottest code paths, we're going to give the "scatter-gather" approach an attempt and test to see how it affects performance in the future. Please add the above index to this PR. Thanks!

mandyvenables replied:
I added a separate index for `tokenBatch.internalId` to `tokenization-tokenBatch`.

dlongley replied:
Thanks!

dlongley commented on lib/tokens.js:
```diff
@@ -54,6 +54,12 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
     fields: {'tokenBatch.id': 1},
     options: {unique: true, background: false}
+  }, {
+    // This does not include the shard key. This is not expected to be ran in
+    // the hottest code paths, so we are taking a "scatter-gather" approach
```
```suggestion
    // this does not include the shard key; it is not expected to be used in
    // the hottest code paths, so a "scatter-gather" approach is acceptable
```

dlongley reviewed (approved):
Ok, approving, but let's hold on merging until its clear it meets some higher level use cases as mentioned before. Thanks!

dmitrizagidulin reviewed (approved):
:+1:

dlongley reviewed (changes requested):
Noting that we need additional changes here to address issues that can arise from partitioned state.

dlongley reviewed (approved):
There may be some loose strings lying around here somewhere, but given that this invents a primitive to directly deal with the issues we had before, I'm ok with moving forward with it. I don't expect there to be significant structural problems as there were before, maybe just some bugs that will come to light with more testing.
